docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,902 @@
+# `MeshTrafficRoute`
+
+- Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4743
+Other issues covered:
+
+- https://github.com/kumahq/kuma/issues/4691
+- https://github.com/kumahq/kuma/issues/4690
+
+## Context and Problem Statement
+
+Now that we have the [new policy matching resources](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/005-policy-matching.md), it's time to create one for traffic routing.
+We currently have both `TrafficRoute` and `MeshGatewayRoute`.
+
+This MADR does not address Gateway API integration per se but does consider Gateway
+API's `HTTPRoute`, both as inspiration as well as to make sure we don't block
+integration.
+
+## Decision Drivers
+
+- `MeshGatewayRoute` and `TrafficRoute` have a large overlap in functionality
+  and the use cases for routing from a gateway and between services are similar
+- Gateway API has different resources per protocol
+
+## Considered Options
+
+- Two resources: `MeshGatewayRoute` and `MeshTrafficRoute`
+- Unified resource for services and gateway: `MeshTrafficRoute`
+- Unified resources but one per protocol, e.g. `MeshHTTPRoute`, `MeshTCPRoute`
+- Two resources: `MeshGatewayRoute` and `MeshTrafficRoute`, which both point to
+  a `HTTPRouteRules` resource
+
+## Decision Outcome
+
+We decided to create two resources `MeshGatewayHTTPRoute` and `MeshHTTPRoute`.
+They have the exact same structure for routing rules but the way they refer to
+targets is different.
+
+Each future protocol will get its own resource:
+
+- `MeshHTTPRoute`
+- `MeshTCPRoute`
+
+which can be extended with others like GRPC.
+
+### `MeshHTTPRoute`
+
+The new resource routes and alters requests from one service to another
+depending on where the request coming from and where it's going to.
+Routes end up in the client-side Envoy config so the source of the affected requests will
+be under `spec.targetRef`. The destination we can select in `spec.to`.
+
+```yaml
+spec:
+ targetRef: ... # from where requests are coming
+ to:
+  - targetRef: ... # to where requests are going
+    rules: ...
+```
+
+The route that's used for a given request is the most specific one that Kuma can
+apply given the `kuma.io/protocol` tag.
+
+### `MeshGatewayHTTPRoute`
+
+This new resource attaches directly to a `MeshGateway` via `spec.targetRef`:
+
+```yaml
+spec:
+ targetRef:
+  kind: MeshGateway # leave open potential other selector kinds
+  name: edge-gateway
+ hostnames:
```

Comment:
Should we put `hostnames` and `rules` under the `default` like we don in `MeshTrace`?

Reply:
`hostnames` is more like a refinement of `targetRef` IMO so I don't think it belongs under `default`. The proposal explains how `default` goes under each `rule` instead of top level.

Reply:
Yeah, I'm talking specifically about the gateway case. Do you expect to have `default` under each `rule` here as well?

Reply:
That was the idea of course, but there doesn't really seem to be a potential conflict between user roles here like there is for a service. Perhaps it's not necessary.